fix: error with consecutive cart processing

g-volker opened the pull request:
- Get extension from original promotion LineItem and assign it to new LineItem

<!--
Thank you for contributing to Shopware! Please fill out this description template to help us to process your pull request.

Please make sure to fulfill our contribution guideline (https://developer.shopware.com/docs/resources/guidelines/code/contribution?category=shopware-platform-dev-en/contribution).

Do your changes need to be mentioned in the documentation?
Please create a second pull request at https://github.com/shopware/docs
-->

### 1. Why is this change necessary?

If a promotion LineItem was added to the cart the extension was not added correctly, because of which an error 
occurred when the processor was called again. 
The LineItem (the discount) was added again because the processor looked for the id
in the extension which wasnt there. All this led to an error when saving the cart, due to the promotion LinteItem  being duplicated.

### 2. What does this change do, exactly?

The PromotionDeliveryCalculator searches the oldCart for the promotion, 
if it already exists it adds the extension to the new LineItem.

### 3. Describe each step to reproduce the issue or behaviour.

Prerequisites: 
1. Create a Promotion which discounts the shipping costs by 100% and set it to active. (Shopping basket rule could be 
    value>100)

Execution:
1. Go to admin page and open order overview page
2. Open existing order or create one with atleast one item and open it
3. Add another item to the cart (should still trigger the promotion)
4. Add another item to the cart (Price will not be calculated)
5. Try saving the order -> error 
```
'Line item with identifier "018faf405d697431beb99530f3145ab0" is not stackable and the quantity cannot be changed.'
```
(The discount Item)


### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfill them.

g-volker committed:
fix: Changed Extension Handling

- Get extension from original promotion LineItem
and assign it to new LineItem

g-volker committed:
fix: Changed Extension Handling

Changelog

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/shopware/shopware?pullRequest=5420) <br/>All committers have signed the CLA.

AydinHassan commented:
hi @g-volker thanks for the PR! Can you add a test for this please?

vienthuong commented on changelog/_unreleased/2024-11-09-Extension-handling-fix.md:
```diff
@@ -0,0 +1,11 @@
+author_github: g-volker
+---
+# CORE
+* Changed the way the PromotionDeliveryCalculator works to fix the extension handling. Retrieves the extension
```
```suggestion
# Core
* Changed the way the PromotionDeliveryCalculator works to fix the extension handling. Retrieves the extension from the old promotion LinteItem and adds it to the new promotion LineItem in order for it to not be created again later.
```

vienthuong commented on changelog/_unreleased/2024-11-09-Extension-handling-fix.md:
````diff
@@ -0,0 +1,11 @@
+# CORE
+* Changed the way the PromotionDeliveryCalculator works to fix the extension handling. Retrieves the extension
+    from the old promotion LinteItem and adds it to the new promotion LineItem in order for it to not be created again later.
+    ```
````
```suggestion
    
```

vienthuong commented:
Hi @g-volker 
Besides a small remark on the changelog, I tried to reproduce the issue but are not able to. Could you describe the step 3 a little bit clearer?

marcelbrode commented:
Hi there! Any updates on this ticket?

marcelbrode commented:
Since we noticed no changes in the last 4 weeks, I'll close this ticket. Feel free to ping me, if you've rebased your branch and provided the changes asked and I will gladly reopen this ticket. Thank you :smile:

g-volker committed:
Merge branch 'shopware:trunk' into fix/extension-handling

g-volker committed:
added ExtensionHandling Unit-Test

g-volker committed:
removed unused quotes

g-volker commented:
Hi @marcelbrode,

i finally got around to updating this pr, please excuse the delay. 
I have updated the description, rebased the branch and created a test for my changes. I guess the pr will update if its reopened, i have already pushed them to my fork.

Thank you for your patience 😄

vienthuong commented:
Hi @g-volker 

We applied your commit and play around with creating new order and the behavior is rather non consitent:

### prerequisites:

-  Create a promotion "TEST PROMOTION" with 100% discount shipping cost, condition is the basket > 100 euro

### Behavior (see the recording):

- TEST PROMOTION is applied correct and shipping fee is deducted
- Add a new product into the basket caused the promotion appears/disappears randomly
- After a few try, there is an error when saving the order

https://github.com/user-attachments/assets/871d04bc-62e4-487d-aea9-78195bad8715

g-volker commented:
Hi @vienthuong,

i've also made a short video showing the behaviour before my changes where the original error occured.
Because of the changes this new bug was uncovered, which is hardly in the scope of this pull request.
The changes in the proccessor work as intented and we should probably consider this a new bug.

https://github.com/user-attachments/assets/116b21b9-b74d-4ddb-b7b1-93601cdee952

How should we/I proceed?

vienthuong commented:
Hi @g-volker 
Thank you for your effort!

Even though the new bug might be not related to your change, I think we need to cover the new uncovered bug anyway.

It'd be great if you can continue with it as you are already familiar with the behavior, or we can continue on your work.

g-volker commented:
Hello all,
thank your for your reply @vienthuong!

Im currently examining the new bug, but since im a little strapped for time it might take a moment for me to fix it.
If i cannot solve it in (reasonable) time ill update you.

g-volker commented:
Hello everyone,

Unfortunately, I haven't been able to resolve this issue, and due to time constraints, I won’t be able to continue working on it. I’d appreciate it if someone else could take over.

Sorry for the late notice, and thank you for takin over in advance!

vienthuong commented:
Hi @g-volker thank you for your work 👍🏽 We will take over it
